server/src/main/java/org/elasticsearch/search/internal/ReaderContext.java

```diff
@@ -0,0 +1,203 @@
+    private final List<Releasable> onCloses = new CopyOnWriteArrayList<>();
+
+    private final long startTimeInNano = System.nanoTime();
+
+    private Map<String, Object> context;
+
+    public ReaderContext(long id,
+                         IndexService indexService,
+                         IndexShard indexShard,
+                         Engine.SearcherSupplier searcherSupplier,
+                         long keepAliveInMillis,
+                         boolean singleSession) {
+        this.id = new SearchContextId(UUIDs.base64UUID(), id);
+        this.indexService = indexService;
+        this.indexShard = indexShard;
+        this.searcherSupplier = searcherSupplier;
+        this.singleSession = singleSession;
+        this.keepAlive = new AtomicLong(keepAliveInMillis);
+        this.lastAccessTime = new AtomicLong(nowInMillis());
+        this.refCounted = new AbstractRefCounted("reader_context") {
+            @Override
+            protected void closeInternal() {
+                doClose();
+            }
+        };
+    }
+
+    private long nowInMillis() {
+        return indexShard.getThreadPool().relativeTimeInMillis();
+    }
+
+    @Override
+    public final void close() {
+        if (closed.compareAndSet(false, true)) {
+            refCounted.decRef();
+        }
+    }
+
+    void doClose() {
+        Releasables.close(Releasables.wrap(onCloses), searcherSupplier);
+    }
+
+    public void addOnClose(Releasable releasable) {
+        onCloses.add(releasable);
+    }
+
+    public SearchContextId id() {
+        return id;
+    }
+
+    public IndexService indexService() {
+        return indexService;
+    }
+
+    public IndexShard indexShard() {
+        return indexShard;
+    }
+
+    public Engine.Searcher acquireSearcher(String source) {
+        return searcherSupplier.acquireSearcher(source);
+    }
+
+    public void keepAlive(long keepAlive) {
+        this.keepAlive.updateAndGet(curr -> Math.max(curr, keepAlive));
+    }
+
+    /**
+     * Marks this reader as being used so its time to live should not be expired.
+     *
+     * @return a releasable to indicate the caller has stopped using this reader
+     */
+    public Releasable markAsUsed() {
+        refCounted.incRef();
+        return Releasables.releaseOnce(() -> {
+            this.lastAccessTime.updateAndGet(curr -> Math.max(curr, nowInMillis()));
+            refCounted.decRef();
+        });
+    }
+
+    public boolean isExpired() {
+        if (refCounted.refCount() > 1) {
+            return false; // being used by markAsUsed
+        }
+        final long elapsed = nowInMillis() - lastAccessTime.get();
+        return elapsed > keepAlive.get();
+    }
+
+    // BWC
+    public ShardSearchRequest getShardSearchRequest(ShardSearchRequest other) {
+        return Objects.requireNonNull(other);
+    }
+
+    public ScrollContext scrollContext() {
+        return null;
+    }
+
+    public AggregatedDfs getAggregatedDfs(AggregatedDfs other) {
```

Comment:
Is it worth to remove `getAggregatedDfs` and `setAggregatedDfs` from  `ReaderContext` and have them only in `LegacyReaderContext`?

Reply:
Yeah, this was the first choice. However, I went with a different approach to avoid too much casting.